config.md

```diff
@@ -384,31 +384,80 @@ For POSIX platforms, the configuration structure supports `hooks` for configurin
+    * **`poststop`** (array of objects, OPTIONAL) is an array of [`poststop` hooks](#poststop).
+        Entries in the array have the same schema as `prestart` entries.
 
 Hooks allow users to specify programs to run before or after various lifecycle events.
 Hooks MUST be called in the listed order.
-Hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
 The [state](runtime.md#state) of the container MUST be passed to hooks over stdin so that they may do work appropriate to the current state of the container.
 
 ### <a name="configHooksPrestart" />Prestart
 
-The pre-start hooks MUST be called after the [`start`](runtime.md#start) operation is called but [before the user-specified program command is executed](runtime.md#lifecycle).
+The `prestart` hooks MUST be called after the [`start`](runtime.md#start) operation is called but [before the user-specified program command is executed](runtime.md#lifecycle).
 On Linux, for example, they are called after the container namespaces are created, so they provide an opportunity to customize the container (e.g. the network namespace could be specified in this hook).
 
+Note: `prestart` hooks were deprecated in favor of `create-runtime`, `create-container` and `start-container` hooks, which allow more granular hook control during the create and start phase.
+`prestart` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
+
+### <a name="configHooksCreateRuntime" />Create Runtime Hooks
+
+The `create-runtime` hooks MUST be called as part of the [`create`](runtime.md#create) operation after the runtime environment has been created (according to the configuration in config.json) but before the `pivot_root` or any equivalent operation has been executed.
+The `create-runtime` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
+
+On Linux, for example, they are called after the container namespaces are created, so they provide an opportunity to customize the container (e.g. the network namespace could be specified in this hook).
+
+Note: `runc` originally implemented `prestart` hooks contrary to the spec, namely as part of the `create` operation (instead of during the `start` operation). This incorrect implementation actually corresponds to `create-runtime` hooks. For runtimes that implement the deprecated `prestart` hooks as `create-runtime` hooks, `create-runtime` hooks MUST be called after the `prestart` hooks.
+
+### <a name="configHooksCreateContainer" />Create Container Hooks
+
+The `create-container` hooks MUST be called as part of the [`create`](runtime.md#create) operation after the runtime environment has been created (according to the configuration in config.json) but before the `pivot_root` or any equivalent operation has been executed.
+The `create-container` hooks MUST be executed in the [container namespace](glossary.md#container-namespace).
```

Comment:
Who ensures this?

Reply:
If the hooks are provided access information such as the linux namespaces for the container, they can join the namespaces and do their work. However at this point, `create-container` and `create-runtime` could be the same.

Reply:
The runtime -- the point is that the runtime should spawn the hook in the container's mount (and other) namespaces *before* performing `pivot_root`.

Reply:
I still don't get `create-container`.

Is the difference between `create-runtime` and `create-container` just that in the latter case all the user namespaces are joined?

Are the `create-container` hooks supposed to be spawned from the container init process (just before `pivot_root`)?  
I guess no, as at this point the container is not yet fully configured and different runtimes may perform the initialization in a different order.  Conversely, if it is called from the runtime process, it just seem like a special case of `create-runtime` where the runtime also joins the namespaces before exec'ing the hook executable.

Reply:
I don't really see why it's better to punt doing `nsenter` off to the hook authors. Given how trivial it is to implement as part of the runtime, and the fact that LXC has incredibly similar hooks, I don't think adding more annoyances to hook authors is a good idea. Especially if your hooks are incredibly trivial bash scripts, having to mess with nsenter just to do something as simple as create a mount seems excessive.

Reply:
> It's not required, but that is one way of doing it (and is probably how we'd implement it in `runc`).

At this point the container is not fully configured, what is left to configure and what is already configured is not known to the hook.  Since there is no requirement on what the OCI runtime must do before the `pivot_root` (beside of course setting the mounts), the hook might run from different contexts depending on the OCI runtime used and the container configuration.  It will easily turn into issues like "hook X works with the runtime Y but doesn't work with the runtime Z".

e.g. should cgroups apply to the hook process?  Seccomp?  Resource limits?

Reply:
Hmm, you have a point. How about we make it `preexec-runtime` and you spawn the hook right before the container's `execve` time (but after the `start` socket has been hit) -- this is kind of the intended use-case for the whole `create`+`start` split but no higher-level runtimes expose that as a feature that you can use for your own extensions (in fact the lack of usability when using higher-level runtimes of the `create`+` start` split is the entire reason for this PR to extend hooks).

But in that case we'd have to ask the hook to deal with joining all of the relevant namespaces -- not to mention that dealing with a container after `pivot_root` isn't really nice if you want to do any operations without touching (**extremely untrusted and dangerous**) container binaries.

Reply:
@RenaudWasTaken, thoughts on this? How is it the equivalent hook used in LXC?

Reply:
> It will easily turn into issues like "hook X works with the runtime Y but doesn't work with the runtime Z".

I expect this to happen over time to be honest, runc, gvisor and kata are already pretty different and need different configs to happen.

> Since there is no requirement on what the OCI runtime must do before the pivot_root (beside of course setting the mounts)

This is a problem that is not just specific to the `create-container` but also to the `create-runtime` hook, though the way it is addressed in spec today is in the `runtime.md` file, step 2 reads as follows:
> container's runtime environment MUST be created according to the configuration in [`config.json`](config.md).  While the resources requested in the [`config.json`](config.md) MUST be created, the user-specified program (from [`process`](config.md#process)) MUST NOT be run at this time. Any updates to [`config.json`](config.md) after this step MUST NOT affect the container.

> @RenaudWasTaken, thoughts on this? How is it the equivalent hook used in LXC?

This would be the `mount hook` equivalent of lxc (before pivot_root and in the container namespace), which is the behavior we are trying to mirror here.

Reply:
@mrunalp @giuseppe @cyphar @RenaudWasTaken
This thread seems unresolved.

1. Should there be a distinction between `create-runtime` and `create-container` at all?
1. If so, how is it this specified?
1. Should hooks be expected to be compatible across different runtimes at all?

To me, the answer to the last question should be yes (that's why there's a runtime-spec), but maybe this goal is at odds with the kind of granular hook control that some applications need...?

Reply:
The answer to the last question most definitely is yes. I think that @giuseppe has a valid point -- this is arguably why we created the whole `create`+`start` split to begin with (it allows you to do hook-like operations without significantly constraining the runtime implementation). But no higher-level runtime exposes the `create`+`start` split properly.

And so we need to use hooks to achieve the same thing. I think that both approaches have downsides:

* Making the hook effectively act as a hook for "post-create" means that the container will have already `pivot_root`ed, making it harder to implement hooks reasonably. There's almost certainly many possible security bugs you could spawn by asking third-party hooks to safely interact with a fully-set-up container rootfs. To be fair, those worries also exist with this PR's hooks -- but by running them before `pivot_root` iti sn't as bad.

* The approach in this PR (as @giuseppe says) results in more and more problems where "runc is the standard". While I would like to avoid this, at this point it's a bit naive to pretend that runc's semantics have not already been implemented by all other runtimes. Maybe the solution here is to just define these semantics in the spec, and move on.

Anyway, I think that (given the security worries I have about "post-create" hooks) we should just go with this PR and see whether the other runtimes have any strong complaints. If nobody screams, then we can just stick with this.

Reply:
Sorry, the past two weeks have been hard for getting some work done on this topic. I'll be more available during December :)

> Should there be a distinction between create-runtime and create-container at all?

This is a nice to have as a hooks author, because that means not having to deal with joining the different namespaces (namely the mount one). As pointed out by @cyphar above it would also help avoid security issues if we mess up our implementation.

> If so, how is it this specified?

I was hoping that by leaving the definition vague and defining an example in the runtime.md file it would be enough.
e.g:

`config.md`:
> The `create-runtime` hooks MUST be called as part of the [`create`](runtime.md#create) operation after the runtime environment has been created (according to the configuration in config.json) but before the `pivot_root` or any equivalent operation has been executed.

`runtime.md`
> 1. OCI compliant runtime's [`create`](runtime.md#create) command is invoked with a reference to the location of the bundle and a unique identifier.
> 2. The container's runtime environment MUST be created according to the configuration in [`config.json`](config.md).
> While the resources requested in the [`config.json`](config.md) MUST be created, the user-specified program (from [`process`](config.md#process)) MUST NOT be run at this time.
> ...
> The [`create-runtime` hooks](config.md#create-runtime-hooks) MUST be invoked by the runtime.


> Should hooks be expected to be compatible across different runtimes at all?

Can you detail what you mean by compatible?

As a hooks developer, I have pretty low expectation of my hook working on multiple runtimes without me having to handle different scenarios (for example we need to pass-through the device in Kata before hand if I want to have access to it in the hook). 
I'm more hoping to have a uniformity of the time it is expected to be called.

To be honest, I don't have a strong opinion of either scenario, I can work with both. If `create-container` is blocking this PR, I'm happy to drop it. It's a nice to have, will likely avoid headaches to many people but we can work around it not being here.

Reply:
> @RenaudWasTaken: Can you detail what you mean by compatible?

I'm really not the best person to argue this case, but the clear goal of a spec (as reaffirmed by @cyphar) above is that different implementations should be compatible, including the hooks. Of course, the different hooks may depend on each other (in the order that they are guaranteed to be executed; c.f. your device example), but the point is that this should work equally across implementations.

A starting point for open questions might be [this comment](https://github.com/opencontainers/runtime-spec/pull/1008#discussion_r347354422) by @giuseppe:

> e.g. should cgroups apply to the hook process? Seccomp? Resource limits?

Reply:
could we perhaps just document that there should not be expectations on the status of the container initialization (such as cgroups/seccomp might be already in place) except that the specified mounts are created?

At least hook developers will be aware of such potential issues so that if a specific configuration is needed another kind of hook can be used?

Reply:
> @giuseppe: could we perhaps just document that there should not be expectations on the status of the container initialization (such as cgroups/seccomp might be already in place) except that the specified mounts are created?

I think this is a good idea - in fact, I'm wondering if this couldn't/shouldn't be strengthened to be mandated by the spec?

I'll admit my understanding is limited, but to me, the question that @giuseppe raised is basically where `create-container` lives relative to the non-mount container setup:

1. `create` command is invoked
1. runtime environment created according to `config.json`
1. `create-runtime` hook is executed
1. mount points are set up, change into container namespace
1. ??? unspecified order of
    - `create-container` hook is executed
    - various bits of container setup: seccomp, cgroups, etc.
1. `pivot_root`
1. `start` command is invoked
1. `start-container` hook is executed
1. ...

I think that we could hit two birds with one stone by prescribing:
1. `create` command is invoked
1. runtime environment created according to `config.json`
1. `create-runtime` hook is executed
1. mount points are set up, change into container namespace
1. `create-container` hook is executed
1. all other container setup: seccomp, cgroups, etc.
1. `pivot_root`
1. `start` command is invoked
1. `start-container` hook is executed
1. ...

The two birds in question are:
* there's a clear specification that's unambiguous (i.e. only the mount and namespace operations happen between `create-runtime` and `create-container`) - which means the answer should be clear (if something doesn't work), whether it's the hook or the runtime that's not spec-conformant. If a hook relies on some of the other bits of container setup, it must be shifted to `start-container`.
* The spec is minimal in the way that `create-container` hooks written for this spec are not at risk of breaking if a later spec version adds something more between the two hooks, because correct hooks (as per the old spec) would/could not be relying on the additional steps.

Reply:
The motivation for the create-runtime hook was to specify the behavior that is currently implemented in runc. That behavior [fully sets up the container (including cgroups, though I haven't looked into selinux)](https://github.com/opencontainers/runc/blob/c8337777b666dec6f4f602849b81096983cdac6a/libcontainer/process_linux.go#L375-L385) before invoking the prestart hooks.

I had in my mind the following flow:
1. `create` command is invoked
2. runtime environment created according to config.json
    mount points are set up, as well as other container setup operations: seccomp, cgroups, etc.
3. `create-runtime` hook is invoked
4. `create-container` hook is invoked 
5. `pivot_root` or equivalent operations
6. `start` command is invoked
7. `start-container` hook is executed

As an example, the `mount` hooks in LXC have joined all the namespaces of the container. We could either follow that model or drop the create-runtime hook and leave the work of joining the different namespace to the hook.

Side note, I'm a bit confused as to why we can't specify completing all container setup operations "except pivot_root or an equivalent operation" before the create-container / create-runtime operation.

Reply:
> Side note, I'm a bit confused as to why we can't specify completing all container setup operations "except pivot_root or an equivalent operation" before the create-container / create-runtime operation.

There are some operations which have to be done as late as possible -- much later than `pivot_root` (such as seccomp application or setting SELinux/AppArmor labels). It's better to not permit containers to `pivot_root`.

Reply:
> @giuseppe: could we perhaps just document that there should not be expectations on the status of the container initialization (such as cgroups/seccomp might be already in place) except that the specified mounts are created?

@cyphar @mrunalp @vbatts 
What do you think about @giuseppe's suggestion? That would probably be the fastest way forward. Or should the spec be more specific?

Reply:
I think to solve this particular blocking problem we should just go with @giuseppe's suggestion to make hook authors aware that this is an underspecified area of the spec and the only guarantees they can be sure of are that the mounts will be set up and `pivot_root` will not have been called yet.

In the future we can work on making the spec more specific.

Reply:
Updated the PR to echo these thoughts :)